### ISSUE TITLE
Carrot from Stackoverflow

### ISSUE THREAD

**Issue by ImCarrot** [derMaaster/GitExamples#1]
Hey! I went over your code, what exactly is the problem? I faced a `CompileTime` issue with your `ToggleSwitchToggled_Command` in the `x:Bind` :

```
<interactcore:InvokeCommandAction Command="{x:Bind ViewModel.ToggleSwitchToggled_Command}" />
```

Is that the issue? 

Because I see the UI is getting populated via `ViewModel` as I mentioned in my answer by defining the `ViewModel` in codebehind as:
````
public OtherPage_ViewModel ViewModel => this.DataContext as OtherPage_ViewModel;
````

**Comment by ImCarrot** [comment]
If that's the issue _(command is not binding)_ then let's talk a bit about why is it happening. It might not be evident until we start using `x:Bind` everywhere. So lets just do that:

------------------------------

I moved the `DataTemplate` to `Page.Resources` to make the two (`DataTemplate` and `DataGridColumn`) seem separated.

````
xmlns:i="using:Microsoft.Xaml.Interactivity"
xmlns:core="using:Microsoft.Xaml.Interactions.Core"

<Page.Resources>
    <DataTemplate x:Key="RunToggleButtonTemplate" x:DataType="vms:VMData">
        <StackPanel>
            <ToggleSwitch IsOn="{x:Bind Running, Mode=TwoWay}" Toggled="RunButton_Toggled">
                <i:Interaction.Behaviors>
                    <core:EventTriggerBehavior EventName="Toggled">
                        <core:InvokeCommandAction Command="{x:Bind ViewModel.ToggleSwitchToggled_Command}" />
                    </core:EventTriggerBehavior>
                </i:Interaction.Behaviors>
            </ToggleSwitch>
        </StackPanel>
    </DataTemplate>
</Page.Resources>
````

and now your `DataGridColumn` looks like:
````
<controls:DataGridTemplateColumn Header="Run" CellTemplate="{StaticResource RunToggleButtonTemplate}" />
````

Now if you notice:
- The `DataContext` of the `DataTemplate` has been fixed to `VMData`. _(Because that's what you were binding it to)_
- The `ToggleSwitch.IsOn`is `x:Bind`ed with `VmData.Running` property. Which works.
- When you try to `x:bind` Command to `ViewModel.ToggleSwitchToggled_Command`, it would not work. _(Compile Time error)_

> The reason is straight forward though, it's trying to find a property called `ViewModel` in your `VMData` class. Which it can't find because it's not there. Remember, Just like `Binding`, `x:Bind` is also binding to a `DataContext`. It just makes it easier to Bind to CodeBehind when we're not using a `DataContext`.

So how do we solve it?
We create a new `RoutedEventHandler` in our `CodeBehind` and bind the `Toggled` event to it. Then simply in CodeBehind call `ViewModel.HandleVmDataRunChanged()` which would do everything. If you want the some relevant data from the `ToggleSwitch` on toggle, then you can use the `DataContext` of that `ToggleSwitch`.

**Comment by ImCarrot** [comment]
### Here is the code

I am only putting in the changes.

In `OtherPage.xaml`:

````
<Page.Resources>
    <DataTemplate x:Key="RunToggleButtonTemplate" x:DataType="vms:VMData">
        <StackPanel>
           <ToggleSwitch IsOn="{x:Bind Running, Mode=TwoWay}" Toggled="RunButton_Toggled" />
        </StackPanel>
    </DataTemplate>
</Page.Resources>

<controls:DataGridTemplateColumn Header="Run" CellTemplate="{StaticResource RunToggleButtonTemplate}" />

````

In the `OtherPage.xaml.cs`:
````
 private void RunButton_Toggled(object sender, RoutedEventArgs e)
{
    if (!(sender is ToggleSwitch selectedToggleSwitch))
        return;

    if (!(selectedToggleSwitch.DataContext is VMData selectedVMData))
        return;

    ViewModel.HandleVmDataRunChanged(selectedVMData);
    // or you can call the ToggleSwitchToggled, if you do then make sure you change the accessibility. 
    // Although I wouldn't recomend passing in an instance of the ToggleSwitch to the ViewModel
    // ViewModel.ToggleSwitchToggled(selectedToggleSwitch);
}
````
in your `OtherPage_ViewModel.xaml.cs`:
````
internal void HandleVmDataRunChanged(VMData selectedVMData)
{
    // Do something here!
}
````

**Comment by derMaaster** [comment]
Hi Aditya,

Thanks for the help thus far. Yes the issue is the command not binding (it
is the same when I use a *ToggleButton* that has *command* built in, where
I do not have to use *microsoft.Xaml.Interactivity*.

However, I do not want to use code behind, and want to strictly stick to
MVVM principles of being loosely coupled. Thus, is there not
another solution where I can use loosely coupled MVVM, with a ViewModel
that houses all my code and xaml events (via command?)

It seems like the problem is, you can't have bindings to two different
dataContext, when you want to use an event/command within a subset of xaml
code?

Regards,
derMaaster

**Comment by ImCarrot** [comment]
That's correct, you can only have a single `DataContext` per control but child elements can only be assigned `DataContext` based on the Parent's `DataContext`. That being said, you can trigger the command via `CodeBehind` but I think if you post another question that targets this particular problem, there might be someone who could help you with this. _(not sure though)_.  As far as this question is concerned, my answer pretty much answers it. 

If you do post another question, do send out the link to me, I am really curious about the answer to this problem.

Regards
Aditya
